### PR TITLE
[Model Loader] Harden startup weight-load overlap and expand model coverage

### PR DESCRIPTION
**Pull request by yh0903**
## Summary

This is the hardening and model-coverage follow-up to #32017. Startup weight-load overlap remains opt-in and `serial` remains the default.

- Add `--startup-weight-load-mode=auto`: admitted configurations use overlap; rejected configurations keep the serial path.
- Centralize exact architecture, native-class resolution, and validation rules in a static profile registry.
- Resolve checkpoint sources before capture-safe mutation, so `auto` can fall back for PT or secondary weights without allocating the model twice.
- Keep graph-visible parameter, buffer, and explicitly registered derived-tensor storage stable across the real-weight commit.
- Cooperatively stop and drain page-cache prefetch before the normal commit. A timed-out worker stays visible to the loader's conservative concurrency policy and is not waited on twice.
- Preserve serial semantics for optional checkpoint parameters and report overlap phases without inflating the legacy `load_weight` metric.

Explicit `overlap` fails with deterministic, actionable rejection reasons. `auto` logs the same reasons and selects serial before model mutation; failures after capture-safe mutation remain terminal.

## Supported profiles

| Profile | Admitted configuration |
| --- | --- |
| Native dense | Native `LlamaForCausalLM`, `Qwen2ForCausalLM`, or `Qwen3ForCausalLM`; FP16/BF16; TP1/TP2 |
| Qwen3.5-family hybrid VLM | Native `Qwen3_5ForConditionalGeneration` (including Qwen3.6 dense checkpoints); BF16; TP2/TP4; Triton linear attention; breakable or disabled prefill CUDA graphs |
| Qwen3.5 MoE hybrid VLM | Native `Qwen3_5MoeForConditionalGeneration`; BF16; TP2+EP2; Triton MoE and linear attention; standard EP; breakable or disabled prefill CUDA graphs |
| Qwen3 MoE EP | Native `Qwen3MoeForCausalLM`; BF16; TP2+EP2; Triton MoE; standard EP (`moe_a2a_backend=none`) |
| GLM-5.2 FP8 | Native `GlmMoeDsaForCausalLM` with the GLM-5.2 fingerprint; dynamic E4M3 FP8 with 128x128 blocks; BF16 execution/KV; Hopper SM90; TP8/TP16, EP1; DSA+FA3; Triton FP8 GEMM/MoE; prefill graphs disabled; shared-expert fusion disabled |

The GLM profile requires the native GLM-5.2 IndexShare fingerprint: default `cli_factor`, no custom `index_topk_pattern`, `index_topk_freq=4`, and `index_skip_topk_offset=3`. The runtime result below covers TP16; TP8 is admission-tested only.

Other architectures, quantized/repacked-weight paths, speculative decoding, LoRA/offload, and backend/topology combinations outside these profiles remain rejected. In `auto` mode they use serial startup.

## How it works

1. **Admit.** Global rules and the selected profile produce an ordered rejection report. Native model implementation is resolved lazily only after cheap preflight succeeds.
2. **Prepare and capture.** The loader initializes the model once, resolves checkpoint sources, fills required parameters with finite sentinels, records graph-visible storage, and prefetches checkpoint pages while KV pools, attention backends, and CUDA graphs initialize.
3. **Commit and validate.** After capture, the manager drains prefetch, restores optional constructor values, loads real weights, reruns post-load processing, and verifies storage identity and sentinel replacement before entering the existing post-load barrier.

Source-format fallback happens before sentinel mutation. Once capture-safe preparation starts, commit or validation failures fail startup instead of retrying serial loading on a mutated model. Background prefetch itself remains best-effort.

### Qwen3.6-27B, BF16 TP2

`Qwen/Qwen3.6-27B@6a9e13bd6fc8f0983b9b99948120bc37f49c13e9`, FA3, Triton GDN/linear attention, prefill CUDA graph disabled, decode FullCG, local NVMe.

Three interleaved trials per mode ran after JIT warmup. Checkpoint pages were evicted and verified non-resident before every measured start.

| Metric (median) | Serial | Overlap | Change |
| --- | ---: | ---: | ---: |
| Process ready | 50.193 s | 46.122 s | **-8.11% / 1.088x** |
| Scheduler startup | 30.03 s | 26.30 s | -12.4% |
| `load_weight` | 11.66 s | 7.50 s | -35.7% |

Every overlap trial used active handoff: capture ended with 8-9 of 15 shards fully prefetched, the worker stopped in about 3.8 s, and multithreaded commit resumed. There were no fallback, timeout, manifest, or sentinel errors. Across all six starts, fixed prompts produced bit-identical text, token IDs, input/output logprobs, and top-logprobs.

The requested decode graph BS 256 resolved to the model's BS 39 capacity in both modes, with capture shapes `[1, 2, 4, 8, 12, 16, 24, 32, 39]`. These timings compare complete `serial` and `overlap` startup paths, including prefetch and handoff.

### GLM-5.2 FP8, TP16

`zai-org/GLM-5.2-FP8@ba978f7d347eaf65d22f1a86833408afdb953541`, BF16 execution/KV, 16 H100s, DSA+FA3, sgl-kernel top-k, DeepGEMM paged MQA, Triton FP8 GEMM/MoE, prefill graph disabled, decode FullCG up to BS 64, local NVMe.

A serial pilot established the compilation-cache baseline. The overlap pilot, six interleaved measured trials, and `auto` validation then started from cold checkpoint pages. Across those eight cold-controlled starts, the maximum resident fraction was `2.71e-8`, and compilation-cache membership and size stayed unchanged.

| Metric | Serial | Overlap | Change |
| --- | ---: | ---: | ---: |
| Process ready (median) | 135.482 s | 125.709 s | **-7.21% / 1.078x** |
| Process ready (mean) | 135.564 s | 125.464 s | -7.45% |
| Cross-trial CV | 0.086% | 0.280% | — |

All five overlap-enabled starts drained active prefetch on all 16 ranks before normal multithreaded commit resumed. All nine starts served successfully and matched text, token IDs, and input/output logprobs for the fixed single and batched probes. There were no fallback, prefetch-timeout, storage-manifest, or sentinel errors. Cross-node coordination-window upper bounds were 1.18-2.39 ms.

### Additional correctness coverage

- `Qwen/Qwen3.5-35B-A3B`, BF16 TP2+EP2: text, image, batched, and prefix-branching requests matched across `serial`, `overlap`, and `auto`.
- `Qwen/Qwen3-30B-A3B-Instruct-2507`, BF16 TP2+EP2: multi-route batched and prefix-branching requests matched across all three modes.

These are correctness/topology probes; the Qwen3.6 and GLM-5.2 experiments above are the performance results.

## Tests

- Focused startup/prefetch suite at the benchmarked product head: `84 passed, 98 subtests passed`.
- The registry refactor matched the previous admission behavior in 10,000 randomized cases, including rejection codes, messages, and order.
- Unit coverage includes profile admission and near misses, deterministic fallback boundaries, storage/sentinel validation, optional parameter restoration, timing ownership, prefetch EOF/cancel/crash/drain races, and loader routing.
- The registered one-GPU test compares `overlap` and `auto` with default serial output IDs, text, and per-token logprobs.
- Changed Python files pass `py_compile`; `git diff --check` passes.

## Follow-up

This PR intentionally keeps `serial` as the default. Follow-ups can add separately validated DeepSeek V4 Flash and Kimi Linear/K3 profiles, define a capture-stable contract for repacking quantization methods, broaden backend/topology evidence, isolate graph overlap with a coordinated-prefetch serial control, and then consider automatic overlap by default.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #35311406427](https://github.com/sgl-project/sglang/actions/runs/35311406427)
Latest PR Test (Extra): :x: [Run #35311406158](https://github.com/sgl-project/sglang/actions/runs/35311406158)
Latest PR Test (AMD ROCm 10): :x: [Run #35311406620](https://github.com/sgl-project/sglang/actions/runs/35311406620)

### CODE REVIEW

**Comment by yh0903** [comment]
@codex review

File: python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py [hunk 1]
```diff
@@ -126,6 +185,34 @@ def from_server_args(
...
+            enable_dp_attention=get_parallel().enable_dp_attention,
+            enable_two_batch_overlap=get_exec().overlap.enable_two_batch_overlap,
+            enable_eplb=get_exec().moe.enable_eplb,
+            ep_num_redundant_experts=get_exec().moe.ep_num_redundant_experts,
+            init_expert_location=get_exec().moe.init_expert_location,
+            elastic_ep_backend=get_exec().moe.elastic_ep_backend,
+            enable_elastic_expert_backup=(
+                get_exec().moe.enable_elastic_expert_backup
+            ),
+            ep_join_mode=get_parallel().ep_join_mode,
```
[the remark this answers is not in the document]
Comment (yh0903) [reply python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py]: Good catch — fixed the lookup to use the exec.moe namespace and added an explicit schema assertion.

File: python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py [hunk 2]
```diff
@@ -141,6 +224,657 @@ def from_published_config(
...
+            "MoE data parallelism is not supported",
+        ),
+        (
+            "moe_a2a_backend",
+            options.moe_a2a_backend != "none",
+            f"{family} startup overlap requires the standard EP path",
+        ),
+        (
+            "moe_runner_backend",
+            options.moe_runner_backend != "triton",
```
Comment (ch-wan) [inline python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py]: **[suggestion]** Profile admission compares published backend *strings* to the validated concrete values (`moe_runner_backend == "triton"`, and for GLM also `fp8_gemm_runner_backend == "triton"` and `dsa_prefill_backend == "fa3"`). Those knobs default to `"auto"` / `None` and are often left unresolved in the published bags: Hopper BF16 MoE still reports `moe_runner_backend="auto"` while `UnquantizedMoeMethod.create_moe_runner` selects Triton, and Hopper BF16 KV DSA defaults `dsa_prefill_backend` to `flashmla_sparse` (decode `fa3`). `--startup-weight-load-mode=auto` therefore serial-falls-back for the advertised Qwen3 MoE / GLM-5.2 profiles unless the user also pins the exact backend flags from the PR table; `--startup-weight-load-mode=overlap` hard-errors with the same mismatch.

**Suggestion:** Admit the resolved runtime backend (treat published `"auto"` as the Triton path on the configs already validated), or resolve DSA/MoE/FP8 GEMM backends before the profile checks so `auto` overlaps the same configs the runtime actually runs.

File: python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py [hunk 3]
```diff
@@ -141,6 +224,657 @@ def from_published_config(
...
+                "Qwen3.5-family hybrid VLM startup overlap requires BF16",
+            ),
+            (
+                "quantization",
+                model_config.quantization is not None,
+                "quantization is not supported",
+            ),
+            (
+                "modelopt",
+                bool(getattr(model_config, "modelopt_quant", False)),
```
Comment (ch-wan) [inline python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py]: **[suggestion]** Every profile repeats `bool(getattr(model_config, "modelopt_quant", False))`. Real `ModelConfig` has no `modelopt_quant` attribute (ModelOpt is `quantization` / `_get_modelopt_quant_type()`); the getattr is always False on production configs and only fires on the unit-test `SimpleNamespace`. The no-getattr rule also applies: this is defensive access for a field that is not part of the type.

**Suggestion:** Drop the getattr and reject on `model_config.quantization` (already covered) or a real ModelConfig field such as `_get_modelopt_quant_type()`.

File: python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py [hunk 4]
```diff
@@ -457,10 +1322,25 @@ def prepare(self) -> nn.Module:
...
+            )
+            model = self._loader.load_initialized_model_from_resolved_sources(
+                model=model,
+                model_config=self._model_config,
+                resolved_sources=resolved_sources,
+                target_device=torch.device(self._device_config.device),
             )
+            self._model = model
+            self._resolved_sources = resolved_sources
+            self._state = StartupWeightLoadState.READY
```
Comment (ch-wan) [inline python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py:901]: **[suggestion]** Auto source-format fallback sets `StartupWeightLoadState.READY` without `_timings`. `finalize()` treats READY as an idempotent successful commit and `assert self._timings is not None`. That is safe today only because `load_model_utils` drops the manager when `not is_deferred`. Any later caller that keeps the manager (metrics, a second finalize, extra runners) will assert instead of no-opping the serial path.

**Suggestion:** Record zero/sentinel timings on the serial fallback READY path, or make `finalize()` a no-op when READY was reached without overlap (no timings) rather than asserting.

File: python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py [hunk 5]
```diff
@@ -245,21 +1007,204 @@ def unchanged_parameter_names(self, value: float) -> Tuple[str, ...]:
...
+                profile=profile,
+                model_config=model_config,
+                options=options,
+            )
+        )
+    else:
+        rejections.append(
+            StartupWeightLoadRejection(
+                code="architecture",
+                message="exactly one supported model architecture is required",
```
Comment (ch-wan) [inline python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py]: **[nit]** A single unsupported architecture (e.g. `MixtralForCausalLM`) and empty/multi-architecture configs share the message "exactly one supported model architecture is required". The old allowlist wording distinguished "wrong arch" from "not exactly one arch"; operators with one unsupported model now get a count error.

**Suggestion:** Keep a separate `architecture` message for "this architecture is not in the overlap registry" versus "need exactly one architecture".

**Review by ch-wan — COMMENTED** [review]
## Summary

This is a careful hardening of opt-in startup weight-load overlap: `auto` admits from a profile registry, source-format fallback happens before sentinel mutation, prefetch is drained before commit (timeout keeps the conservative single-reader policy), optional `_skip_weight_check` parameters are restored for serial post-load semantics, and GLM MLA derived tensors are included in the storage-identity check. Routing now follows the actual manager rather than `--startup-weight-load-mode=overlap`, which is the right auto-mode contract, and ModelRunner stays orchestration-only. Dominant remaining risk is admission using unresolved backend *strings* (`auto` / `None`) instead of the backends the runtime actually selects, so advertised Qwen3 MoE and GLM profiles will not overlap under default flags.

## Issue counts by severity

- bugs: 0
- suggestions: 3
- nits: 1

File: python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py (hunk 2 above)
Comment (yh0903) [reply python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py]: Thanks for the suggestion. I've now removed the per-profile backend allowlists. Normal backend selection stays in the existing runtime; admission now focuses on overlap-specific loading/storage constraints.

File: python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py (hunk 3 above)
Comment (yh0903) [reply python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py]: Thanks! Just removed this check and the test-only field, and quantization checks now use `model_config.quantization` and checkpoint metadata.

File: python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py (hunk 4 above)
Comment (yh0903) [reply python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py:901]: Fixed: finalize() returns None after serial fallback. The caller preserves serial timing and avoids repeating the barrier. Both paths have regression coverage.

File: python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py (hunk 5 above)
Comment (yh0903) [reply python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py]: Removed the single-architecture restriction in the latest commits. We now follow native registry selection and report when the selected implementation has no overlap loading path.

File: python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py [hunk 6]
```diff
@@ -70,77 +96,375 @@ class StartupWeightLoadState(str, enum.Enum):
...
+) -> Tuple[Tuple[str, bool, str], ...]:
+    from sglang.srt.layers.moe.utils import (
+        MoeA2ABackend,
+        resolve_moe_runner_backend,
+    )
+
+    runner = resolve_moe_runner_backend(options.moe_runner_backend)
+    rules = (
+        (
+            "lora",
```
Comment (ch-wan) [inline python/sglang/srt/model_executor/model_runner_components/startup_weight_load.py:280]: **[bug]** LoRA (and UNO) is rejected only inside `_moe_weight_rules`. Native dense and Qwen3.5 hybrid VLM use `_validate_native_dense`, which never consults `options.has_lora`, so `--enable-lora` / UNO + overlap is admitted for Llama/Qwen/Qwen3.5, including serialized block-FP8. `ModelRunner.initialize()` still wraps modules in `maybe_init_lora_manager()` after `load_model()` (sentinel prepare) and before CUDA graph capture / `finalize_startup_weight_load()`. `BaseLayerWithLoRA` re-exposes only `.weight` and `.bias` on the wrapper; FP8 `weight_scale` / `weight_scale_inv` / `input_scale` stay under `*.base_layer.*`. `load_weights()` looks up the unwrapped names, skips the scales, and `unchanged_parameter_names` then raises after graphs are already captured. Auto cannot fall back at that point (`commit failures are terminal`). Unquantized dense LoRA may still work via the `.weight` alias; the hole is the LoRA x non-weight-parameter (quantized) cross product this PR newly admits.

**Suggestion:** Treat `options.has_lora` as a global overlap rejection (same rule previously used for every architecture), or delay LoRA wrapping until after a successful commit. If dense unquantized LoRA is intentionally supported, still reject `has_lora` whenever `_checkpoint_block_fp8_size` is non-empty (and any other path whose checkpoint names are not aliased onto the wrapper). Add a unit admission case for native_dense/Qwen3.5 + LoRA + block-FP8 so auto selects serial instead of crashing after capture.

File: python/sglang/srt/layers/radix_attention.py [hunk 7]
```diff
@@ -154,6 +154,11 @@ def __init__(
         self.logit_capping_method = logit_capping_method
         self.xai_temperature_len = -1
 
+    def named_startup_weight_load_constants(self):
```
Comment (ch-wan) [inline python/sglang/srt/layers/radix_attention.py:157]: **[suggestion]** `k_scale_float` / `v_scale_float` are now graph-visible constants. Capture-time KV postprocess on optional `_skip_weight_check` parameters produces the default `1.0`; a later checkpoint scale updates the Python floats and `ModelStorageManifest.changed_names` aborts commit. That fail-loud is correct for forced overlap, but `auto` admits `kv_cache_dtype=fp8_e4m3` (explicitly not an admission input) and only falls back before sentinel mutation. Models whose checkpoints actually contain KV scales will still pay for graph capture and then raise `RuntimeError` instead of loading serially.

**Suggestion:** Keep the constant check. For `auto`, either reject FP8 KV-cache configs at admission, or document that non-default checkpoint KV scales are a late, terminal overlap failure. Forced `--startup-weight-load-mode=overlap` can stay fail-loud.
